.circleci/real_config.yml

```diff
@@ -112,85 +112,74 @@ commands:
-            MERGE_BASE="$(git merge-base upstream/main HEAD)"
-            DIFF_DIRS="$(git diff-tree --no-commit-id --name-only "$MERGE_BASE" HEAD)"
-            if [ "$DIFF_DIRS" = "docs" ] ; then
-              echo "docs-only change detected, halting job now"
-              circleci-agent step halt
-            else
-              echo "detected non-docs-only changes: $DIFF_DIRS"
-            fi
+
+  skip-if-dir:
+    parameters:
+      dir:
+        type: string
+      not:
+        type: boolean
+        default: false
+    steps:
+      - unless:
+          condition: << parameters.dir >>
+          steps:
+            - run:
+                name: skip if << parameters.dir >>-only changes
+                command: |
+                  MERGE_BASE="$(git merge-base upstream/main HEAD)"
+                  DIFF_DIRS="$(git diff-tree --no-commit-id --name-only "$MERGE_BASE" HEAD)"
+                  if [ "$DIFF_DIRS" = "<< parameters.dir >>" ] ; then
+                    echo "<< parameters.dir >>-only change detected, halting job now"
+                    circleci-agent step halt
+                  else
+                    echo "detected non-<< parameters.dir >>-only changes: $DIFF_DIRS"
+                  fi
+      - when:
+          condition: << parameters.dir >>
+          steps:
+            - run:
+                name: skip if not << parameters.dir >> changes
+                command: |
+                  MERGE_BASE="$(git merge-base upstream/main HEAD)"
+                  DIFF_DIRS="$(git diff-tree --no-commit-id --name-only "$MERGE_BASE" HEAD)"
+                  if [[ "$DIFF_DIRS" == *"<< parameters.dir >>"* ]]; then
+                    echo "<< parameters.dir >> change detected, running downstream steps"
+                    circleci-agent step halt
+                  else
+                    echo "<< parameters.dir >> code not affected, halting job now"
+                  fi
+
+  skip-if-docs-only:
+    steps:
+      - skip-if-dir:
+          dir: docs
 
   skip-if-not-docs:
     steps:
-      - run: git remote add upstream https://github.com/determined-ai/determined
-      - run: tools/scripts/retry.sh git fetch upstream
-      - run:
-          name: check for any docs changes
-          command: |
-            MERGE_BASE="$(git merge-base upstream/main HEAD)"
-            DIFF_DIRS="$(git diff-tree --no-commit-id --name-only "$MERGE_BASE" HEAD)"
-            # check "shared" is in the list of changed directories
-            if [[ "$DIFF_DIRS" == *"docs"* ]]; then
-              echo "docs change detected, running downstream steps"
-            else
-              echo "detected non-docs changes, halting job now"
-              circleci-agent step halt
-            fi
+      - skip-if-dir:
+          dir: docs
+          not: true
 
   skip-if-webui-only:
     steps:
-      - run: tools/scripts/retry.sh git fetch upstream
-      - run:
-          name: check for webui-only changes # must be run after skip-if-docs-only
```

Comment:
this comment is what made me want to move `add-and-fetch-upstream` to it's own command